### PR TITLE
feat(typeahead): add support for custom templates for search results

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
cc @maxokorokov

### CODE REVIEW

File: src/typeahead/typeahead-window.ts [hunk 1]
```diff
@@ -47,4 +64,6 @@ export class NgbTypeaheadWindow {
   prev() { this.activeIdx = (this.activeIdx === 0 ? this.results.length - 1 : this.activeIdx - 1); }
 
   select(item) { this.selectEvent.emit(item); }
+
+  private prepareTplCtx(result: any) { return new ResultTplCtx(result, this.term, this.formatter); }
```
Comment (maxokorokov) [inline src/typeahead/typeahead-window.ts]: underscore -> `_prepareTplCtx`
and also -> `_activeIdx` by the way

File: src/typeahead/typeahead-window.ts [hunk 2]
```diff
@@ -37,8 +45,17 @@ export class NgbTypeaheadWindow {
    * A function used to format a given result before display. This function should return a formated string without any
    * HTML markup.
    */
-  @Input() formatter = (result) => { return `${result}`; };
+  @Input() formatter = toStringFormatter;
```
Comment (maxokorokov) [inline src/typeahead/typeahead-window.ts]: just curious, why replace this?
maybe put it to utils, as we're using it elswhere?

**Comment by maxokorokov** [comment]
Looks alright to me!

File: src/typeahead/typeahead-window.ts (hunk 2 above)
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead-window.ts]: > just curious, why replace this?

Oh, this is due to TS / linter. TS interprets `formatter = (result) => { return `${result}`; };` as a "method" member and our linting rules are forcing methods after regular fields. Put the 2 together and you have to have `formatter` as the last field (even after outputs!).

> maybe put it to utils, as we're using it elswhere?

Makes sense. Will move it in a separate PR.

Comment (pkozlowski-opensource) [inline src/typeahead/typeahead-window.ts]: @maxokorokov here: https://github.com/ng-bootstrap/ng-bootstrap/pull/464
